Schemas-API.adoc

```diff
@@ -0,0 +1,64 @@
+# Proposal: Schema Manipulation in Hotrod for Protostream Infinispan
+
+## Goal
+To enhance schema evolution and management in Infinispan's Protostream serialization framework, we propose replacing the current API with a new schema manipulation mechanism.
+
+## Hotrod client side marshaller
+
+### Client side api today
+
+[source, java]
+----
+Schema schema = // programmatic schema
+MagazineMarshaller magazineMarshaller = new MagazineMarshaller(); // custom marshaller 1
+BookMarshaller bookMarshaller = new BookMarshaller(); // custom marshaller 2
+
+ProtoStreamMarshaller protostreamMarshaller = new ProtoStreamMarshaller();
+SerializationContext serializationContext = marshaller.getSerializationContext();
+FileDescriptorSource fds = FileDescriptorSource.fromString(schema.getName(), schema.toString());
+serializationContext.registerProtoFiles(fds);
+serializationContext.registerMarshaller(magazineMarshaller);
+serializationContext.registerMarshaller(bookMarshaller);
+builder.marshaller(protostreamMarshaller);
+builder.remoteCache("cacheName").marshaller(protostreamMarshaller);
+----
+
+### Client side api proposal
+
+[source, java]
+----
+Schema schema = // programmatic schema
+MagazineMarshaller magazineMarshaller = new MagazineMarshaller(); // custom marshaller 1
+BookMarshaller bookMarshaller = new BookMarshaller(); // custom marshaller 2
+
+ProtoStreamMarshaller protostreamMarshaller = new ProtoStreamMarshaller();
+protostreamMarshaller.register(schema, magazineMarshaller, bookMarshaller);
+
+builder.remoteCache("cacheName").marshaller(marshaller);
+----
+
+### Administration API
+
+[source, java]
+----
+public interface RemoteCacheManagerAdmin {
+    SchemasAdministration schemas();
+}
+----
+
+[source, java]
+----
+public interface SchemasAdministration {
+    CompletionStage<SchemaOpResult> createAsync(Schema schema);
+    SchemaOpResult create(Schema schema);
+
+    CompletionStage<SchemaOpResult> updateAsync(Schema schema);
+    SchemaOpResult update(Schema schema);
+
+    CompletionStage<SchemaOpResult> deleteAsync(String name);
+    SchemaOpResult delete(String name);
```

Comment:
Hmmm, delete seems quite dangerous. If the schema is removed, the data in a cache would only be accessible as `byte[]` afterward. What would be the use cases for removing a schema? I understand it exists in the current API as well, but maybe we could hide some dangerous operations from users.

Reply:
It would technically make any cache that has their encoding set to protostream no longer accessible even as it couldn't deserialize the values. You would also have to update any cache to be octet-stream or unknown encoding iirc.